src/engine/streaming_engine_perdevice.cc

```diff
@@ -0,0 +1,1257 @@
+
+    int noCores;
+
+    LazyAllocArray<TaskRunner<kFIFO          >> cpu_normal_workers_;   ///< cpu worker
+    std::unique_ptr<TaskRunner<kPriority     >> cpu_priority_worker_;  ///< cpu priority worker
+    LazyAllocArray<StreamRunner               > gpu_stream_runners_;   ///< stream runner for GPU
+
+    template<typename T>
+    int randomGPUStream(const T &block) {
+        std::uniform_int_distribution<int> distribution(0, gpu_streams_ - 1);
+        return distribution(block->generator);
+    }
+
+    StreamRunner *getStreamRunner(int dev_id) {
+        auto msr  = [this, dev_id]() { return GetStreamRunner(dev_id); };
+        return gpu_stream_runners_.Get(dev_id, msr);
+    }
+
+    /// Detector for GPU events signaling operations have finished
+    /// \param dev_id The device id of the worker.
+    /// \param block The task block of the worker.
+    /// \param stream id to detect events on
+    inline void EventWorker(int dev_id,
+                            StreamRunner *block,
+                            int si) {
+#       if MXNET_USE_CUDA
+        string threadName = string("Event Worker thread") + std::to_string(si);
+        NVTX_NAMETHREAD((uint32_t)(size_t) (pthread_self()), threadName.c_str());
+
+        // Wait for events on queues until exit
+        mshadow::SetDevice<gpu>(dev_id);
+
+        vector<shared_ptr<Operation>> ops;
+        shared_ptr<Operation> op;
+        while (block->event_queues[si]->Pop(&op)) {
+            ops.clear();
+            ops.push_back(op);
+
+            for (int i = 0; i < 10; i++) {  // Burst
+                if (block->event_queues[si]->ApproxSize() > 0) {
+                    block->event_queues[si]->Pop(&op);
+                    ops.push_back(op);
+                } else { break; }
+            }
+
+            for (int i = 0; i < ops.size(); i++) {
+                shared_ptr<Operation> &op = ops[i];
+
+//                NVTX_RANGE_END();
+#               if MXNET_USE_NVTX
+                string range_name = string("opsToBeScheduled ") + op->name +
+                                    std::to_string(static_cast<int>(static_cast<size_t>op->opr));
+#               endif
+                NVTX_RANGE_START(rangeName.c_str() , 0x00008800);
+
+                // Launch further ops if possible
+                vector<shared_ptr<Operation>> scheduled = op->opsToBeScheduled(Operation::Launched);
+                NVTX_RANGE_END();
+
+                NVTX_RANGE_START("PushToExecute", 0x00008800);
+
+//                NVTX_RANGE_START("PushToExecute", 0x00008800);
+                if (scheduled.size() > 0) {
+                    for (auto &&o : scheduled) this->PushToExecute(o, false);
+                }
+//                NVTX_RANGE_END();
+
+                NVTX_RANGE_END();
+            }
+
+            // Wait for completion
+            auto &&stream = mshadow::Stream<gpu>::GetStream(block->streams[si].get());
+            MSHADOW_CUDA_CALL(cudaStreamSynchronize(stream));
```

Comment:
In this approach, all synchronization is done on separate threads (one per stream) so that CPU/memory operations and GPU operations with multiple dependencies can be scheduled when dependencies have finished (has synchronized). The constant 10 you see (should be refactored) is just for bursting through multiple GPU ops pushed onto the synchronization queue as an optimization - it allow us to effectively skip some unnecessary synchronization. In the existing engines synchronization is done after *all* operations.
I hope to explain it all better when I'm back from holidays. More benchmarking is of course a good thing regardless!